doc/user-manual/language/type-based-termination-checking.rst

```diff
@@ -0,0 +1,317 @@
+   fib .head = zero
+   fib .tail .head = suc zero
+   fib .tail .tail = zipWith plus fib (fib .tail)
+
+This function passes termination checking. We shall explain the logic of Agda for the third clause.
+
+Following our intuition with coinductive functions, the are three depth parameters ``k < m < n``, where the outer stream is of depth ``n``, and to pass checking the third clause should return the stream of depth at least ``k``. If the first inner ``fib`` is used with the depth ``k``, and the second ``fib`` is used with the depth ``m`` (note, that the smallest available depth for ``fib .tail`` is ``k``, hence ``fib`` must use something bigger, which is ``m``), then the size-preserving ``zipWith`` returns a stream of size ``k``, which is indeed what is required from it. Now we see that both recursive calls to ``fib`` are performed with depths ``k`` and ``m``, which are smaller than ``n``. Agda concludes that this function is terminating.
+
+Size preservation is tightly coupled with polarities. Given a function signature, all occurrences of *negative inductive* and *positive coinductive* variables are considered as input, and they serve as possible candidates for size preservation analysis. On the other hand, all *positive inductive* and *negative coinductive* variables are considered as output, and a function signature may be size-preserving precisely in them. For example, consider the following definition:
+
+::
+
+    foo : {A : Set} → (Nat → A) → Nat → A
+    foo f x = f x
+
+
+
+Here, the first ``Nat`` in ``foo`` is in a doubly-negative position, which means that the position is positive, and ``foo`` can be size-preserving in the first ``Nat``. From the definition we see that it is indeed the case. One application of this fact is that the following function passes termination check:
+
+::
+
+    bar : Nat → Nat
+    bar zero = zero
+    bar (suc n) = foo bar n
+
+As a final note, we address performance considerations. Currently, size-preservation analysis is the slowest part of the type-based termination checker. If you suspect that it causes a slowdown, you can specify ``--no-size-preservation``, disabling the analysis while retaining the rest of the type-based termination checker. Nevertheless, there are plans to improve its performance.
+
+.. _type-based-termination-checking-size-limitations:
+
+Limitations
+-----------
+
+The most significant limitation of the current implementation is rooted in the fact that the size type system relies on System Fω, while the target language of Agda is dependently-typed. In cases where a type signature of a function involves large elimination, it is likely that the type-based termination checker will encounter difficulties. This limitation arises because dependent types introduce additional complexity to the underlying theory, which was initially developed for a variant of System Fω. Further details on the semantical framework can be explored in :ref:`[3]<type-based-termination-checking-references>`.
```

Comment:
Could you give an example here of a situation where this problem actually shows up?

Reply:
Isn't mentioning large elimination enough? We already have an instance of this in the standard library, the reflection modules abuse large elimination, and I cannot check termination of them because of this fact.